src/markdown_exec/formatters/jupyter.py

```diff
@@ -0,0 +1,253 @@
+                    ):
+                        if shell_msg["content"]["traceback"][0] == "Stack trace:":
+                            error_outputs.extend(
+                                shell_msg["content"]["traceback"][1:],
+                            )
+                        else:
+                            error_outputs.extend(shell_msg["content"]["traceback"])
+                    break
+                # note: abort is deprecated, but some older kernels may still send it
+                # it doesn't send error info, however - so we'll need to populate a fake abort error instead
+                if shell_msg["content"]["status"] == "abort":
+                    error_outputs.append("AbortError: Execution aborted")
+                    break
+        except queue.Empty:
+            pass
+
+        try:
+            # nested while here so we consume all iopub messages for this execution before moving on.
+            # this lets us consume all pending messages after execution has completed before exiting the loop
+            while True:
+                # Check for iopub messages (output)
+                iopub_msg = kc.get_iopub_msg(timeout=0.1)
+                if iopub_msg["parent_header"].get("msg_id") != msg_id:
+                    continue
+
+                msg_type = iopub_msg["header"]["msg_type"]
+                content = iopub_msg["content"]
+
+                if msg_type == "stream":
+                    outputs.append(content["text"])
+                elif msg_type in ("execute_result", "display_data"):
+                    # TODO: these can contain other MIME types (including HTML) - do we want to support them?
+                    outputs.append(str(content["data"].get("text/plain", "")))
+                elif msg_type == "error":
+                    # ignore these because we already got it from the shell message
+                    pass
+
+        except queue.Empty:
+            # No messages received in timeout period, check if kernel is still alive
+            if not km.is_alive():
+                raise ExecutionError(
+                    "Error: Kernel died during execution",
+                ) from None
+            continue
+
+    # Check for errors
+    if error_outputs:
+        if returncode:
+            return "\n".join(error_outputs)
+        raise ExecutionError(
+            code_block(language, "\n".join(error_outputs), **extra),
+        )
+
+    return "\n".join(outputs)
+
+
+
+def _jupyter_formatter(language: str, runner: Callable[[str], str] | None = None, **kwargs: Any) -> str:
+    """Format and execute code using Jupyter kernels."""
+    return base_format(
+        language=language,
+        run=runner or functools.partial(_run_jupyter, language, {}),
+        **kwargs,
+    )
+
+# if _shutdown_kernels isn't called before the process exits, the windows build
+# of deno panics see https://github.com/denoland/deno/issues/28004 for more info
+atexit.register(_shutdown_kernels)
```

Comment:
lint: I've had a bad experience with projects registering exit functions at the module-level. Please only do so if needed (if relevant part of the code is entered at runtime, not import time).